Add show r code to main ui of teal

averissimo opened the pull request:
# Pull Request

- Fixes https://github.com/insightsengineering/teal/issues/1534
- Companion of:
  - https://github.com/insightsengineering/teal/pull/1592
  - https://github.com/insightsengineering/teal.widgets/pull/319
  - https://github.com/insightsengineering/teal.reporter/pull/401

### Changes description

- Adds "Show R code" to main teal UI
- Supports `disable_show_r_code()` function that is applied to modules to hide it
- Adds support to hide "+ Add to Report" without affecting this new functionality
  - It previously changes the return object to `NULL` fully disabling any post-module functionality
- Adds new global option to disable the "Show R Code" button
  - `+ Add to Report` doesn't need it as it can be disabled in `init`

<img width="1019" height="241" alt="image" src="https://github.com/user-attachments/assets/b3cfa17c-b9a6-4879-8cb5-b004ca816d46" />

### Example app

```r
# Show R Code PR example app

# Make sure the required branches are checked-out
# - teal_reporter: 1534-show_r_code@main
# - teal: 1534-show_r_code@main
pkgload::load_all("../teal.logger")
pkgload::load_all("../teal.code")
pkgload::load_all("../teal.data")
pkgload::load_all("../teal.widgets")
pkgload::load_all("../teal.reporter")
pkgload::load_all("../teal.slice")
pkgload::load_all("../teal.transform")
pkgload::load_all("../teal")

library(teal.logger)
library(teal.code)
library(teal.data)
library(teal.reporter)
library(teal.slice)
library(teal.widgets)
library(teal.transform)
library(teal)

do.call(
  options,
  rlang::list2(
    "teal.bs_theme" = bslib::bs_theme(version = "5"),
    teal.log_level = "TRACE",
    teal.show_js_log = TRUE,
    teal.show_r_code = TRUE,
  )
)

example_extended <- function(label = "example teal module",
                             datanames = "all",
                             transformators = list(),
                             decorators = list()) {
  checkmate::assert_string(label)
  checkmate::assert_list(decorators, "teal_transform_module")

  mod <- example_module(label, datanames, transformators, decorators)

  module(
    label,
    server = function(id, data, decorators) {
      moduleServer(id, function(input, output, session) {
        result <- mod$server("example", data, decorators)

        reactive({
          q <- result()
          attr(q, "teal.enable_src") <- FALSE
          q
        })
      })
    },
    ui = function(id, decorators) mod$ui(shiny::NS(id, "example"), decorators),
    ui_args = mod$ui_args,
    server_args = mod$server_args,
    datanames = mod$datanames,
    transformators = mod$transformators
  )
}

teal::init(
  data = within(teal_data(), {
    iris <- iris
    require(nestcolor)
    CO2 <- CO2
  }),
  modules = modules(
    example_module(label = "Module (full)"),
    example_extended(label = "📓 Module (manual)"),
    example_module(label = "📓 Module (pipe disable show r code)") |> 
      disable_src(),
    example_module(label = "🗒️ Module (pipe disable report)") |> 
      disable_report(),
    example_module(label = "🗒️📓 Module (pipe disable both)") |> 
      disable_src() |> 
      disable_report()
  )
) |> shiny::runApp()
```

m7pr commented on R/source_code.R:
```diff
@@ -0,0 +1,85 @@
+    })
+
+    is_button_showed_r <- shiny::reactive({
+      getOption("teal.show_r_code_ui", TRUE) &&
+        inherits(mod_out_r(), "qenv") &&
+        !isFALSE(attr(mod_out_r()@code, "teal.show_r_code")) # Only hide when value is explicitly FALSE
```
Just double checking, so there is a global option set via options called `teal.show_r_code_ui` to disable globally, and a local attribute of the code `teal.show_r_code` to disable locally?

It would be the best if we could get similar behavior for both Show R Code and Add to the Reporter.

Maybe just:
- disable_report/disable_show_r_code for local disable
- init(reporter = NULL / init(show_r_code = FALSE) for global disable
?

And then forget about `attr(mod_out_r()@code, "teal.show_r_code")`?

averissimo replied:
I'm not completely opposed to adding it to `init`, it just feels a very small option to deserve first-class access to that function

m7pr commented on R/source_code.R:
```diff
@@ -0,0 +1,85 @@
+#'   shinyApp(app$ui, app$server)
+#' }
+#' @export
+disable_src <- function(x) {
+  checkmate::assert_multi_class(x, c("teal_module", "teal_modules"))
+  after(x, server = function(data) {
+    attr(data@code, "teal.show_src") <- FALSE
+    data
+  })
+}
```
@averissimo should this be done as `after`? so it's dispatched for teal_modules that lapply over a dispatch for teal_module like suggested in here https://github.com/insightsengineering/teal/pull/1592#discussion_r2338768547?

Or we don't need to implement `disable_src` like this, because `after` already handles the dispatches?

averissimo replied:
The idea is that `after` already handles the dispatch so that we can create "simpler" functions that contain the functionality

llrs-roche commented:
I was using this PR to prepare a vignette on teal and I noticed that the button on the Encoding panel is still shown:

<img width="500" height="249" alt="image" src="https://github.com/user-attachments/assets/d9798e70-3161-45e1-85aa-6922a4b97d2b" />

I guess this button will be automatically removed or it will require modifications on all modules? Sorry I didn't saw any comment/feedback on this.

averissimo commented on inst/css/custom.css:
```diff
@@ -134,8 +134,14 @@ body > div:has(~ #shiny-modal-wrapper .blur_background) {
-  display: flex;
-  justify-content: flex-end;
-  padding: 10px;
+.outline-button.primary {
```
Agree, I'm modifying

averissimo replied:
I'm convinced we need the rule, as such I made the rule more specific to `teal`.

Hardcoded `:hover:not(:disabled)` is not possible though.

If we want to have some consistent UI AND for each package to stand on its own, we need this repetition. Same reasoing to why `disable-busy` is repeated..

gogonzo commented on R/source_code.R:
```diff
@@ -0,0 +1,85 @@
+        !isFALSE(attr(mod_out_r()@code, "teal.show_src")) # Only hide when value is explicitly FALSE
+    })
+
+    .call_once_when(!is.null(doc_out()), {
```
do we really need `call_once_when`?

gogonzo commented on R/source_code.R:
```diff
@@ -0,0 +1,85 @@
+      script = "busy-disable.js"
+    ),
+    shiny::actionButton(
+      shiny::NS(id, shiny::NS("source_code", "button")),
```
Nice work-around to avoid `verbatim_popup_ui` but not accepting. This is not documented nor commented at all. Because of this `srv_source_code` is confusing where `ui_source_button(session$ns(NULL))` is linked with `teal.widgets::verbatim_popup_srv(id = "source_code")`.

You can consider renaming  `ui_source_button` -> `verbatim_popup_ui` and call it with the same namespace in the renderUI


```suggestion
      shiny::NS(id, "button"),
```

averissimo replied:
Good call, see the new changes in 4bb8ecd42

It "overrides" the `verbatim_popup_ui`

averissimo commented on R/source_code.R:
```diff
@@ -0,0 +1,114 @@
+disable_src <- function(x) {
+  checkmate::assert_multi_class(x, c("teal_module", "teal_modules"))
+  after(x, server = function(data) {
+    attr(data, "teal.enable_src") <- FALSE
```
Discussion for @insightsengineering/nest-core-dev (from coredev corner)

Using attributes set on the S4 object instead of the slots. The attribute names uses a namespace `prefix` to indicate the context.

In the future we may use `shared.xxxx` to pass options to other modules (think `mdr` composite modules). This makes it easier to maintain over time.

averissimo commented on R/source_code.R:
```diff
@@ -0,0 +1,85 @@
+    })
+    observeEvent(doc_out(), ignoreNULL = FALSE, {
+      shinyjs::toggleState("source_code_container", condition = !is.null(doc_out()))
+    })
```
The problem with moving it to `renderUI` is that it re-renders it _(graying out the UI while shiny is computing)_.

Plus. This `toggleState` doesn't control the visibility of the element, it disables/enables the button _(disables when module has error)_

averissimo commented on R/source_code.R:
```diff
@@ -0,0 +1,122 @@
+      shinyjs::toggleState(
+        "source_code_wrapper",
+        condition = !is.null(code_out()) &&
+          checkmate::test_multi_class(mod_out_r(), c("qenv", "qenv.error")) &&
```
Allowing `qenv.error` that has an attribute that disables functionality

averissimo commented on R/source_code.R:
```diff
@@ -0,0 +1,85 @@
+
+    is_button_showed_r <- shiny::reactive({
+      getOption("teal.show_src", TRUE) &&
+        inherits(mod_out_r(), "qenv") &&
```
This may be ignored if we keep the buttons always showing in the UI

averissimo commented on R/module_nested_tabs.R:
```diff
@@ -225,9 +225,16 @@ srv_teal_module <- function(id,
-    ui_add_reporter(ns("add_reporter_wrapper")),
     tagList(
-      .modules_breadcrumb(modules),
+      tags$div(
+        style = "display: flex; justify-content: space-between; align-items: center; width: 100%;",
+        .modules_breadcrumb(modules),
+        tags$div(
+          style = "display: flex; gap: 0.5em;",
+          ui_add_reporter(ns("add_reporter_wrapper")),
+          ui_source_code(ns("source_code_wrapper"))
+        )
+      ),
```
@kumamiao @donyunardi your opinion is welcome here as this is a product feature.

We can implement any of the choices.

kumamiao replied:
Thank you @averissimo for the proposal. I personally prefer the "Keep & disable" option, which encourages our users to utilize more of these features, and makes it more consistent like you said. A few other considerations points:

1. With our recent refactors, it has become much easier for customized modules to enable these features. So, adding them in newly created customized modules should not place too much extra burden on developers.
2. My only concern is with legacy modules, which may have been built with the old customization approach. Updating those may not be straightforward - or the developer may simply not have time. In such cases, app users might question why a feature is “disabled” and request it to be enabled, which could lead to complaints from module developers. I suggest changing the wording from "is disabled" to "is not enabled".
4. Regarding your second point on the global option: will it still display the two buttons but keep them greyed out, or will it completely hide them across the app? If it’s the latter, how does the implementation look like for app developers, similar to disabling them on the module level or via options?

kumamiao replied:
> Do you think this deals with most legacy modules? I'm worried with some that have Show R Code in the encoding panel and then disabled in the top-right

I have a similar concern now. My naive intuition before was that their existing SRC will be directly translated to the SRC on the top-right, but it looks like with the new approach, we no longer have the argument below in `teal.widgets::standard_layout()`, so there will be scenarios like you mentioned - the SRC is still in the encoding, but the upper right showing SRC is not enabled. Any good proposals here? Could there be some utility function to make the conversion? Do we have some documentation on making the updates from legacy modules? Maybe the wording could add something like"..., or has not been updated with the new approach of creating SRC"? 
```
forms = tagList( teal.widgets::verbatim_popup_ui(ns("rcode"), button_label = "Show R code") )
```

Re: implementation, I think the proposal makes sense for it to be not a part of `init`, just wanted to make sure that they're not implemented the same way as the module level, which could lead to confusion.

averissimo commented on R/source_code.R:
```diff
@@ -0,0 +1,122 @@
+
+#' @noRd
+ui_source_button <- function(id, title = NULL) {
+  verbatim_popup_ui(shiny::NS(id, "source_code"), title)
```
I think we should avoid repeated code for the same purpose on the framework. Once that the modules no longer use it, we could eventually move that to `teal`.

I've improved on the implementation.

Alternatively, I could make the necessary change on `teal.widgets` and avoid this UI override.

gogonzo replied:
> Alternatively, I could make the necessary change on teal.widgets and avoid this UI override.


Makes more sense, there is a `disabled` argument (`verbatim_popup_srv(disabled)`), what if the "reason" was implemented in `verbatim_popup_srv`? 
Maybe we can sneak-in reason somehow?

averissimo commented on R/teal_reporter.R:
```diff
@@ -129,20 +127,54 @@ srv_add_reporter <- function(id, module_out, reporter) {
-      shinyjs::toggleState("reporter_add_container", condition = inherits(doc_out(), "teal_card"))
+      reason <- trimws(reason_r() %||% "")
+      if (is.null(reason) || identical(reason, "")) {
+        session$sendCustomMessage("disable-tooltip", session$ns("report_add_wrapper"))
```
If we go with the `renderUI` approach, then the `enabled`/`disabled` toggle, must also be controlled there as `shinyjs` will not be consistent

gogonzo replied:
Got it, what about this https://github.com/insightsengineering/teal/pull/1592/files#r2382268043

gogonzo commented on R/module_nested_tabs.R:
```diff
@@ -225,9 +225,16 @@ srv_teal_module <- function(id,
+        tags$div(
+          style = "display: flex; gap: 0.5em;",
+          ui_add_reporter(ns("add_reporter_wrapper")),
+          ui_source_code(ns("source_code_wrapper"))
```
@averissimo since button is always there, and you don't want to link it with the "reason", maybe it should look like this?

```suggestion
          if (getOption("teal.show_src", TRUE)) {
            bslib::tooltip(
              id = ns("source_code_tooltip"),
              trigger = shiny::tags$div(
                id = session$ns("source_code_wrapper"),
                class = "cursor-helper",
                ui_source_button(session$ns(NULL))
              ),
              uiOutput(ns("reason"))
          )
```

averissimo replied:
Ok, let's make it shiny based with:

The visual glitch is very hard to see and this avoids the observer and the extra JS code to handle `bslib` limitation

```r
  output$source_code_container <- shiny::renderUI({
      if (getOption("teal.show_src", TRUE)) {
        reason <- trimws(reason_r() %||% "")
        new_ui <- verbatim_popup_ui(session$ns("source_code"))
        if (!identical(reason, "")) {
          new_ui <- bslib::tooltip(
            id = session$ns("source_code_tooltip"),
            trigger = shinyjs::disabled(
              shiny::tags$div(
                id = session$ns("source_code_wrapper"),
                class = "cursor-helper",
                new_ui
              )
            ),
            reason
          )
        }
        new_ui
      }
    })
```

averissimo commented:
@gogonzo I've updated the PRs and added https://github.com/insightsengineering/teal.widgets/pull/319

- Static UI function _(only renders the reason for failure or description of button)_
- Render button moved to `teal.widgets` and is shared between `reporter` and `teal`

averissimo commented on vignettes/teal-options.Rmd:
````diff
@@ -156,6 +156,12 @@ list(
 )
 ```
 
+### `teal.show_src` (`logical`)
````
This name is deliberately different from the property to disable inside a module as it has a different goal (hides/removes it)

gogonzo reviewed (approved):
👍
